### PR TITLE
retry, client: support to log after execing some times

### PR DESCRIPTION
**Pull request by CabinfeverB**
### What problem does this PR solve?

Issue Number: ref #7894

### What is changed and how does it work?

```commit-message
```

### Check List

Tests

- Unit test

Code changes

- Has the configuration change
- Has HTTP API interfaces changed (Don't forget to [add the declarative for the new API](https://github.com/tikv/pd/blob/master/docs/development.md#updating-api-documentation))
- Has persistent data change

Side effects

- Possible performance regression
- Increased code complexity
- Breaking backward compatibility

Related changes

- PR to update [`pingcap/docs`](https://github.com/pingcap/docs)/[`pingcap/docs-cn`](https://github.com/pingcap/docs-cn):
- PR to update [`pingcap/tiup`](https://github.com/pingcap/tiup):
- Need to cherry-pick to the release branch

### Release note

```release-note
None.
```

### CODE REVIEW

File: client/retry/backoff.go [hunk 1]
```diff
@@ -16,12 +16,27 @@ package retry
...
 	"github.com/pingcap/failpoint"
+	"github.com/pingcap/log"
+	"go.uber.org/zap"
 )
 
+// Option is used to customize the backoffer.
+type Option func(*Backoffer)
+
+// WithLogTimes sets the number of retries required to print a log.
+func WithLogTimes(logTimes int) Option {
```
Comment (rleungx) [inline client/retry/backoff.go:34]: I think it should belong to the log component?

Comment (CabinfeverB) [reply client/retry/backoff.go:34]: But the log component doesn't know how many times the backoffer has done?
How about providing a function param that the caller defines? like func `WithActionAfterTimes(times int, fn func()) Option`

Comment (rleungx) [reply client/retry/backoff.go:34]: For the caller, I think it's different to know how to set this value, it's affected by the backoff policy. If we must do it, I prefer something like withMinLogInterval.

**Review by rleungx — APPROVED** [review]

File: client/retry/backoff_test.go [hunk 2]
```diff
@@ -119,13 +119,28 @@ func TestBackofferWithLog(t *testing.T) {
...
+	bo := InitialBackoffer(time.Millisecond*10, time.Millisecond*100, time.Millisecond*1000, withMinLogInterval(time.Millisecond*100))
 	err := bo.Exec(ctx, testFn)
 	re.ErrorIs(err, errTest)
 
-	m1 := lg.Message()
-	rfc := `["call PD API failed and retrying"] [api=testFn] [retry-time=10] [error=test]`
-	re.Contains(m1, rfc)
+	ms := lg.Messages()
+	len1 := len(ms)
+	re.Len(ms, 9)
```
Comment (HuSharp) [inline client/retry/backoff_test.go:129]: Can you add more information to explain why care 5 and 13?(such as (1+2+4+8)%10=5)

File: client/retry/backoff.go [hunk 3]
```diff
@@ -16,15 +16,30 @@ package retry
...
 	"go.uber.org/multierr"
+	"go.uber.org/zap"
 )
 
 const maxRecordErrorCount = 20
 
+// Option is used to customize the backoffer.
+type Option func(*Backoffer)
+
+// withMinLogInterval sets the mininum log interval for retrying.
```
[the remark this answers is not in the document]
Comment (HuSharp) [reply client/retry/backoff.go:36]: maybe can add more comments/change this function name, It seems unrelated to the **minimum**?

File: client/retry/backoff_test.go [hunk 4]
```diff
@@ -107,3 +110,91 @@ func TestBackoffer(t *testing.T) {
...
+	re.Contains(ms[len1], rfc)
+}
+
+var errTest = errors.New("test")
+
+func testFn() error {
+	return errTest
+}
+
+// testingWriter is a WriteSyncer that writes to the the messages.
```
Comment (HuSharp) [inline client/retry/backoff_test.go:153]:
```suggestion
// testingWriter is a WriteSyncer that writes the messages.
```

**Review by HuSharp — APPROVED** [review]

**Pull request merged**